sql/mysql_db/jwt_auth.go

```diff
@@ -0,0 +1,240 @@
+
+	var claims jwt.Claims
+	claimsError := fmt.Errorf("ValidateJWT: KeyID: %v. Err: %w", keyID, errKeyNotFound)
+	for _, key := range keys {
+		claimsError = parsed.Claims(key.Key, &claims)
+		if claimsError == nil {
+			break
+		}
+	}
+	if claimsError != nil {
+		return false, claimsError
+	}
+
+	// Verify that the signed JWT is not expired and does not violate "nbf"
+	// Verify all claims on the JWT as configured in the USER and the JWKS.
+	expectedClaims := getExpectedClaims(jwksConfig.Claims)
+	if err := claims.Validate(expectedClaims.WithTime(time.Now())); err != nil {
+		return false, err
+	}
+
+	logString := "Authenticating with JWT: "
+	for _, field := range jwksConfig.FieldsToLog {
+		logString = logString + fmt.Sprintf("%s: %s,", field, getClaimFromKey(claims, field))
+	}
+	logrus.Info(logString)
+	return true, nil
+}
+
+func getClaimFromKey(claims jwt.Claims, field string) string {
+	switch field {
+	case "id":
+		return claims.ID
+	case "iss":
+		return claims.Issuer
+	case "sub":
+		return claims.Subject
+	}
+	return ""
+}
+
+func getExpectedClaims(claims map[string]string) jwt.Expected {
+	var expectedClaims jwt.Expected
+	for cl, value := range claims {
+		switch cl {
+		case "iss":
+			expectedClaims.Issuer = value
+		case "sub":
+			expectedClaims.Subject = value
+		case "aud":
+			expectedClaims.Audience = []string{value}
+		}
+	}
+	return expectedClaims
+}
+
+func getMatchingJwksConfig(config []JwksConfig, name string) (*JwksConfig, error) {
+	for _, item := range config {
+		if item.Name == name {
+			return &item, nil
+		}
+	}
+	return nil, fmt.Errorf("ValidateJWT: Matching JWKS config not found")
+}
+
+func getJWKSFromFile(filepath string) (*jose.JSONWebKeySet, error) {
+
+	file, err := os.Open(filepath)
+	if err != nil || file == nil {
+		return nil, errFileNotFound
+	}
+	defer file.Close()
+
+	byteValue, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	return unmarshalAndConvertKeys(byteValue)
+}
+
+func getJWKSFromUrl(locationUrl string) (*jose.JSONWebKeySet, error) {
+	client := &http.Client{}
+
+	request, err := http.NewRequest("GET", locationUrl, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	} else if response.StatusCode/100 != 2 {
+		return nil, errors.New("FetchedJWKS: Non-2xx status code from JWKS fetch")
+	} else {
+		defer response.Body.Close()
+		contents, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		return unmarshalAndConvertKeys(contents)
+	}
+}
+
+type jsonKey struct {
+	ID  string `json:"id"`
+	Key string `json:"key_bytes"`
+}
+
+func unmarshalAndConvertKeys(byteValue []byte) (*jose.JSONWebKeySet, error) {
+	jsonKeys := []jsonKey{}
+	err := json.Unmarshal(byteValue, &jsonKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make([]jose.JSONWebKey, len(jsonKeys))
+	for i, jKey := range jsonKeys {
+		priv, err := decodePrivateKey(jKey.Key)
```

Comment:
This doesn't seem right. hostedapi is serving an actual []jose.JSONWebKey with public keys. This should be able to use jose methods to parse it, etc. In particular, the test data for this code should be the public JWKS, not the private one that hostedapi and friends are using to generate JWTs.